### PR TITLE
feat(Button/react): add component

### PR DESCRIPTION
**Pull request by krishnaacc123**
feat(Button/react): add component

### CODE REVIEW

File: src/atoms/Button/Button.web.js [hunk 1]
```diff
@@ -1,20 +1,351 @@
...
+            variantColor={variantColor}
+            disabled={disabled}
+            size={size}
+            block={block}
+            variant={variant}
+            {...automation('ds-button')}
+          >
+            <React.Fragment>
+              {icon && iconAlign === 'left' && (
+                <Icon
```
Comment (shubhnik) [inline src/atoms/Button/Button.web.js:285]: The icon seems to have a different color than the text in stories.
Screenshot 2020-04-13 at 12 29 00 PM

Comment (krishnaacc123) [reply src/atoms/Button/Button.web.js:285]: Its because Icon component is not working properly

Comment (krishnaacc123) [reply src/atoms/Button/Button.web.js:285]: I've raised a PR for that.

Comment (krishnaacc123) [reply src/atoms/Button/Button.web.js:285]: https://github.com/razorpay/blade/pull/97

Comment (krishnaacc123) [reply src/atoms/Button/Button.web.js:285]: merged

File: src/atoms/Button/Button.web.js [hunk 2]
```diff
@@ -1,20 +1,351 @@
...
+import Flex from '../Flex';
+import Size from '../Size';
+import Space from '../Space';
+import Icon from '../Icon';
+import Text from '../Text';
+import View from '../View';
+import automation from '../../_helpers/automation-attributes';
+import { getColor, makePxValue, getVariantColorKeys } from '../../_helpers/theme';
+
+const styles = {
```
Comment (shubhnik) [inline src/atoms/Button/Button.web.js:14]: There are also `focus` and `hover` states for Button on the web. Refer Figma.

**Comment by deorukhkarchaitanya** [comment]
No hover or active states?

File: src/atoms/Button/Button.web.js (hunk 2 above)
Comment (krishnaacc123) [reply src/atoms/Button/Button.web.js:14]: Added!

**Comment by krishnaacc123** [comment]
I've added the states

> No hover or active states?

File: src/atoms/Button/Button.stories.js [hunk 3]
```diff
@@ -32,38 +33,46 @@ storiesOf('Button', module)
...
-          <View>
-            <Button
-              size={select('Sizes', sizes, 'medium')}
-              variant={select('Variants', variants, 'primary')}
-              block
-            >
-              Block Button
-            </Button>
-          </View>
+          <Space margin={[0.5, 0.5]} padding={[0.5, 0.5]}>
```
Comment (dhruvdutt) [inline src/atoms/Button/Button.stories.js:36]:
```suggestion
          <Space margin={[0.5]} padding={[0.5]}>
```

File: src/atoms/Button/Button.stories.js [hunk 4]
```diff
@@ -32,38 +33,46 @@ storiesOf('Button', module)
...
+            <View>
+              <Button
+                size={select('Sizes', sizes, 'medium')}
+                variant={select('Variants', variants, 'primary')}
+                block
+              >
+                Block Button
```
Comment (dhruvdutt) [inline src/atoms/Button/Button.stories.js:43]: Move stories with block button below the non-block ones.

File: src/atoms/Button/Button.stories.js [hunk 5]
```diff
@@ -32,38 +33,46 @@ storiesOf('Button', module)
...
-          </View>
+          <Space margin={[0.5, 0.5]} padding={[0.5, 0.5]}>
+            <View>
+              <Button
+                size={select('Sizes', sizes, 'medium')}
+                variant={select('Variants', variants, 'primary')}
+                icon="info"
+                block
+              >
+                Block Button
```
Comment (dhruvdutt) [inline src/atoms/Button/Button.stories.js:57]: This story looks redundant. Remove this. The next one already covers this.

File: src/atoms/Button/Button.stories.js [hunk 6]
```diff
@@ -32,38 +33,46 @@ storiesOf('Button', module)
...
         <Flex flexDirection="row">
-          <View>
-            <Button
-              size={select('Sizes', sizes, 'medium')}
-              variant={select('Variants', variants, 'primary')}
-              icon="info"
-              block
-            />
-          </View>
+          <Space margin={[0.5, 0.5]} padding={[0.5, 0.5]}>
```
Comment (dhruvdutt) [inline src/atoms/Button/Button.stories.js:49]:
```suggestion
          <Space margin={[0.5]} padding={[0.5]}>
```

File: src/atoms/Button/Button.web.js [hunk 7]
```diff
@@ -1,20 +1,454 @@
...
+        return 'primary.800';
+      case 'tertiary':
+        if (disabled) {
+          return 'shade.940';
+        }
+        if (variantColor) {
+          return `${variantColor}.800`;
+        }
+        return 'primary.800';
+      default:
```
Comment (dhruvdutt) [inline src/atoms/Button/Button.web.js:38]: This one is not needed. Default variant is `primary` anyway.

File: src/atoms/Button/Button.web.js [hunk 8]
```diff
@@ -1,20 +1,454 @@
...
+        return '0px';
+    }
+  },
+  height({ size, theme }) {
+    switch (size) {
+      case 'xsmall':
+        return theme.spacings.xlarge;
+      case 'small':
+        return makePxValue(3.5); //TODO: not on scale of 8
```
Comment (dhruvdutt) [inline src/atoms/Button/Button.web.js:176]: What's the issue here? How much is the height in absolute px?

File: src/atoms/Button/Button.web.js [hunk 9]
```diff
@@ -1,20 +1,454 @@
...
+  },
+};
+
+const StyledButton = styled.button`
+  &&& {
+    background-color: ${styles.backgroundColor};
+    border-radius: ${(props) => props.theme.spacings.xxsmall};
+    border: ${styles.border};
+    width: ${(props) => (props.block ? '100%' : '')};
+    outline: none;
```
Comment (dhruvdutt) [inline src/atoms/Button/Button.web.js:328]: Add `cursor: pointer;`

Comment (dhruvdutt) [reply src/atoms/Button/Button.web.js:328]: Add `pointer-events` property, set as `none` for `disabled` state.

File: src/atoms/Button/Button.web.js [hunk 10]
```diff
@@ -1,20 +1,454 @@
...
+  disabled,
+  variantColor,
+  icon,
+  iconAlign,
+  align,
+  block,
+}) => {
+  const theme = useContext(ThemeContext);
+
+  const color = `${variantColor || 'primary'}`;
```
Comment (dhruvdutt) [inline src/atoms/Button/Button.web.js:358]: What's the difference between `color` and `variantColor`? Getting confused.

File: src/atoms/Button/Button.web.js [hunk 11]
```diff
@@ -1,20 +1,454 @@
...
+            onClick={onClick}
+            variantColor={variantColor}
+            disabled={disabled}
+            size={size}
+            block={block}
+            variant={variant}
+            {...automation('ds-button')}
+          >
+            <React.Fragment>
+              {icon && iconAlign === 'left' && (
```
Comment (dhruvdutt) [inline src/atoms/Button/Button.web.js:387]:
```suggestion
              {icon && iconAlign === 'left' ? (
```

Comment (dhruvdutt) [reply src/atoms/Button/Button.web.js:387]: Please don't use `&&` for conditional rendering. Use `&&` for checks and `?` for conditional render. Make change at other places as well.

File: src/atoms/Button/Button.web.js [hunk 12]
```diff
@@ -1,20 +1,454 @@
...
+                <Icon
+                  name={icon}
+                  size={styles.iconSize({ size, children })}
+                  fill={styles.fontColor({ variant, variantColor, disabled })}
+                />
+              )}
+              {icon && iconAlign === 'left' && children && (
```
Comment (dhruvdutt) [inline src/atoms/Button/Button.web.js:394]:
```suggestion
              {icon && iconAlign === 'left' && children ? (
```

File: src/atoms/Button/Button.web.js [hunk 13]
```diff
@@ -1,20 +1,454 @@
...
+                  size={styles.fontSize({ size })}
+                  align="center"
+                  _weight="bold"
+                  _letterSpacing={styles.letterSpacing({ size })}
+                  _lineHeight={styles.lineHeight({ size })}
+                >
+                  {size === 'xsmall' ? children.toUpperCase() : children}
+                </Text>
+              ) : null}
+              {icon && iconAlign === 'right' && children && (
```
Comment (dhruvdutt) [inline src/atoms/Button/Button.web.js:411]:
```suggestion
              {icon && iconAlign === 'right' && children ? (
```

File: src/atoms/Button/Button.web.js [hunk 14]
```diff
@@ -1,20 +1,454 @@
...
+                <Space margin={styles.spaceBetween({ size, iconAlign })}>
+                  <View />
+                </Space>
+              )}
+              {icon && iconAlign === 'right' && (
```
Comment (dhruvdutt) [inline src/atoms/Button/Button.web.js:416]:
```suggestion
              {icon && iconAlign === 'right' ? (
```

File: src/atoms/Button/Button.stories.js (hunk 4 above)
Comment (krishnaacc123) [reply src/atoms/Button/Button.stories.js]: done

File: src/atoms/Button/Button.web.js (hunk 14 above)
Comment (krishnaacc123) [reply src/atoms/Button/Button.web.js]: added

File: src/atoms/Button/Button.web.js (hunk 13 above)
Comment (krishnaacc123) [reply src/atoms/Button/Button.web.js]: added

File: src/atoms/Button/Button.web.js (hunk 12 above)
Comment (krishnaacc123) [reply src/atoms/Button/Button.web.js]: added

File: src/atoms/Button/Button.web.js (hunk 11 above)
Comment (krishnaacc123) [reply src/atoms/Button/Button.web.js]: added

File: src/atoms/Button/Button.web.js (hunk 10 above)
Comment (krishnaacc123) [reply src/atoms/Button/Button.web.js:360]: The consumer can use their own `variantColor` as well. It is a prop on button whereas color is just an alias for it being used in many places.

Comment (krishnaacc123) [reply src/atoms/Button/Button.web.js:360]: I've removed all other instances of `variantColor` so it won;t be much confusing now.

File: src/atoms/Button/Button.web.js (hunk 9 above)
Comment (krishnaacc123) [reply src/atoms/Button/Button.web.js:330]: added

File: src/atoms/Button/Button.web.js (hunk 8 above)
Comment (krishnaacc123) [reply src/atoms/Button/Button.web.js]: removed by confirming with figma

File: src/atoms/Button/Button.stories.js [hunk 15]
```diff
@@ -32,38 +33,46 @@ storiesOf('Button', module)
...
-              size={select('Sizes', sizes, 'medium')}
-              variant={select('Variants', variants, 'primary')}
-              icon="info"
-              block
-              iconAlign={select('iconAlign', iconAlign, 'left')}
-            >
-              Block Button
-            </Button>
-          </View>
+          <Space margin={[0.5, 0.5]} padding={[0.5, 0.5]}>
```
[the remark this answers is not in the document]
Comment (krishnaacc123) [reply src/atoms/Button/Button.stories.js]: added

File: src/atoms/Button/Button.stories.js (hunk 6 above)
Comment (krishnaacc123) [reply src/atoms/Button/Button.stories.js]: added

File: src/atoms/Button/Button.stories.js (hunk 5 above)
Comment (krishnaacc123) [reply src/atoms/Button/Button.stories.js]: removed

File: src/atoms/Button/Button.stories.js (hunk 4 above)
Comment (krishnaacc123) [reply src/atoms/Button/Button.stories.js]: done

File: src/atoms/Button/Button.stories.js (hunk 3 above)
Comment (krishnaacc123) [reply src/atoms/Button/Button.stories.js]: added

File: src/atoms/Button/Button.web.js (hunk 9 above)
Comment (dhruvdutt) [reply src/atoms/Button/Button.web.js:330]: Noice. 💯

Comment (dhruvdutt) [reply src/atoms/Button/Button.web.js:330]: Actually, for `cursor` as well the designers might want to not have `pointer`.
Can you please check once on Figma or someone from the design team?

File: src/atoms/Button/Button.web.js (hunk 8 above)
Comment (dhruvdutt) [reply src/atoms/Button/Button.web.js]: @shubhnik Can you also please check and remove these comments from the Button.native.js as well? Thanks!

File: src/atoms/Button/Button.web.js (hunk 11 above)
Comment (dhruvdutt) [reply src/atoms/Button/Button.web.js]: Noice, thanks!

File: src/atoms/Button/Button.web.js (hunk 10 above)
Comment (dhruvdutt) [reply src/atoms/Button/Button.web.js:360]: I see this `color` approach in native implementation as well. @shubhnik Any specific reason for doing this for native?

File: src/atoms/Button/Button.web.js (hunk 9 above)
Comment (krishnaacc123) [reply src/atoms/Button/Button.web.js:330]: confirmed from design.

File: src/atoms/Button/Button.web.js (hunk 10 above)
Comment (krishnaacc123) [reply src/atoms/Button/Button.web.js]: Makes sense. Replaced it.

**Comment by dhruvdutt** [comment]
Parking this aside for some time. // @kamleshchandnani 
Merged into Shield component [#85](https://github.com/razorpay/frontend-commander/pull/85).

**Comment by dhruvdutt** [comment]
We'll try to merge this. For things not to get blocked, we've cloned this component into shield for now.
Will clean that up once this is merged.

**Pull request closed without merging**